lib/os/mempool.c

```diff
@@ -236,67 +231,72 @@ static void *block_break(struct sys_mem_pool_base *p, void *block, int l,
 int z_sys_mem_pool_block_alloc(struct sys_mem_pool_base *p, size_t size,
 			      u32_t *level_p, u32_t *block_p, void **data_p)
 {
-	int i, from_l, alloc_l = -1, free_l = -1;
-	unsigned int key;
-	void *data = NULL;
+	/* Array of sizes for each level. Populated on demand to save RAM. */
 	size_t lsizes[LVL_ARRAY_SZ(p->n_levels)];
-
-	/* Walk down through levels, finding the one from which we
-	 * want to allocate and the smallest one with a free entry
-	 * from which we can split an allocation if needed.  Along the
-	 * way, we populate an array of sizes for each level so we
-	 * don't need to waste RAM storing it.
-	 */
 	lsizes[0] = _ALIGN4(p->max_sz);
-	for (i = 0; i < p->n_levels; i++) {
+
+	/* Find level from which we want to allocate. */
+	int alloc_l = -1;
+	for (int i = 0; i < p->n_levels; i++) {
 		if (i > 0) {
-			lsizes[i] = _ALIGN4(lsizes[i-1] / 4);
+			lsizes[i] = _ALIGN4(lsizes[i - 1] / 4);
 		}
 
 		if (lsizes[i] < size) {
 			break;
 		}
 
 		alloc_l = i;
-		if (!level_empty(p, i)) {
-			free_l = i;
+	}
+
+	if (alloc_l < 0) {
+		*data_p = NULL;
+		return -ENOMEM;
+	}
+
+	void *data = NULL;
+
+	unsigned int key = pool_irq_lock(p);
+
+	/* Try allocating the smallest block possible.
+	 * Search must be performed atomically to make sure block rearranging
+	 * on layers done by another context will not hide memory from us.
+	 */
+	int from_l;
+	for (from_l = alloc_l; from_l >= 0; from_l--) {
+		data = block_alloc(p, from_l, lsizes[from_l]);
```

Comment:
This is the latency bug: the old code was lockless here, but the new code can potentially call block_alloc() once for every level in the allocator before releasing the lock (for reference, a 32k heap of 16 byte blocks has 8 levels).  That's a lot of cycles added to worst case latency!

Reply:
I have made some measurements today to check the impact.
I write to GPIO just after lock is taken and just before lock released (not including the internal release/lock done on break). This should measure worst case scenario under the lock and show the additional time spent.

I noticed that previous organization of code gave slightly better timing so I stick with it (comments are not updated yet as there is no point if change is not approved).

I used logic analyzer with 25 MS/s so resolution should be ok.
Testing code does `k_malloc`/`k_free` of given size. As nothing is allocated from the heap before the test this is the worst case scenario where blocks will have to be broken down on higher levels.
Heap was configured as follows: 64B - 1024B (default config in our project). I can reconfigure it to have more levels but the pattern is visible so there may be no point.

This was performed on board with nrf52840.

```
size (B) |  base (us)  | change (us)
-------- | ----------- | -----------
      16 | 8.12 - 8.64 | 8.36 - 8.92
      64 | 5.00 - 5.52 | 5.12 - 5.72
     512 | 1.72 - 2.00 us | 1.72 - 2.08
```

The impact seems to be:
- none if block is available on level
- around 0.12 us per level (block alloc call finds level empty)

Cost of 2x gpio write should be around 0.8 us or so.

Cost of block allocation: around 1 us
Cost of single block break: around 3.2 us

I noticed that current code relax the lock after `block_break` which makes latency sum of both times. I moved the relax before break so now the maximum latency is max of block alloc and break times.
In proposed change we have N checks on level being empty added to normal `block_alloc` time. To have max latency on the same level as `block_break` number of levels would have to be around `(3.2 - 1)/0.12` that is around 18.

Reply:
I have increased heap and checked if timings are as expected on higher levels.
```
   level |   base (us)  |  change (us)
-------- | ------------ | ------------
       4 | 14.32 - 14.88 | 14.8 - 15.24
       3 | 11.2 - 11.76 | 11.6 -  12.04
       2 |  8.12 -  8.64 |  8.36 -  8.92
       1 |  5.00 -  5.52 |  5.12 -  5.72
       0 |  1.72 -  2.00 |  1.72 -  2.08
```
It seems to be scaling as expected.

Reply:
Hi @andyross , have you had a chance to look at the values? The worst case scenario is not increased and I would prefer to leave implementation as is to avoid further problems in this area.
If you are fine I will update the comments.